Fix unexpect lead number missing

sfc9982 opened the pull request:
I found the program unexpectly outputs `-` when input `2`. Which should be `+- (+3 + -1 = 2)`

The reason why is that, in pervious code, if `i = 0`, `++ternary[i - 1]` couldn't take affect.   So my solution is manully insert a zero element, though it's slow - **O(n)**.

You can choose to reverse the whole vector at first, and then the addition will be happened on right side, which means you can use `push_back()`.

sfc9982 committed:
Fix unexpect lead number missing

In pervious code, if `i = 0`, ++ternary[i - 1] couldn't take affect.  
So my solution is manully insert a zero element, though it's slow - O(n).
You can choose to reverse the whole vector at first, and then the addition will be happened on right side, which means you can use `push_back()`.

BrandonPacewic commented:
~Shouldn't the behavior of outputting a `-` be expected for the input `2`? I thought that for balanced ternary that was correct as instead of outputting numbers between `0, 1, 2` its between `-1, 0, 1`. Do you have a different example that you could give me?~

BrandonPacewic commented:
Ah I see my error now. Is there possibly a solution other than using `.insert` or `push_back`? As you said, particularly `.insert`, can be quite slow.

sfc9982 commented:
Maybe using a single variable to store the leading number, but it will make the code not that beautiful. I will notice you if I have some better idea.

sfc9982 commented:
Oh, you may iterate the vector from `1` instead of `0`. I will update the code.

sfc9982 committed:
Change the direction of vector

I redirected the way digits stored in vector.

In the usual implementation, ternary numbers are represented using vectors, with each int representing one ternary bit of the number. 

When data being read into a vector, the highest bit of the number is at the head of the vector (the position with the smallest subscript). The reason for this is that **the length of ternary numbers may change** , but we want the same weight bits to remain aligned (e.g., we want all the digits to be in subscript [0], all the tens to be in subscript [1]...); at the same time, we want all the tens to be in subscript [1]... ...); and the fact that addition, subtraction, and multiplication are generally performed from the first digit (think of vertical arithmetic in elementary school) gives a good reason for "inverted storage". And this makes benifits for further numeric computations in this project.

sfc9982 committed:
clang-format edit

When `DerivePointerAlignment` enabled. `PointerAlignment` is then used only as fallback, which breaks the style.

sfc9982 commented on src/balanced_ternary_converter.cpp:
```diff
@@ -28,25 +28,23 @@ std::vector<int> make_ternary(const std::vector<int>& values, int value) {
-            }
-        }
+        ternary[i] += value / values[i];
+        value %= values[i];
```
modular (%) actually done works above

BrandonPacewic replied:
This is a much better solution.

sfc9982 commented on src/balanced_ternary_converter.cpp:
```diff
@@ -28,25 +28,23 @@ std::vector<int> make_ternary(const std::vector<int>& values, int value) {
+        value %= values[i];
     }
+    std::reverse(ternary.begin(), ternary.end());
 
```
I redirected the way digits stored in vector.

In the usual implementation, ternary numbers are represented using vectors, with each int representing one ternary bit of the number. 

When data being read into a vector, the highest bit of the number is at the head of the vector (the position with the smallest subscript). The reason for this is that **the length of ternary numbers may change** , but we want the same weight bits to remain aligned (e.g., we want all the digits to be in subscript [0], all the three-s to be in subscript [1]...); and the fact that addition, subtraction, and multiplication are generally performed from the first digit (think of vertical arithmetic in elementary school) gives a good reason for "inverted storage". 

And this kind of "align" makes benifits for further numeric computations in this project.

sfc9982 replied:
It's similar to high-precision quadrature operations in competitive programming.

BrandonPacewic replied:
I had a feeling that this would have been a better solution from the start. Not sure how I never caught this problem.

sfc9982 commented on src/balanced_ternary_converter.cpp:
```diff
@@ -28,25 +28,23 @@ std::vector<int> make_ternary(const std::vector<int>& values, int value) {
-    for (int i = ternary.size() - 1; i >= 0; --i) {
+    for (int i = 0; i < ternary.size(); ++i) {
+        if (ternary[i] > 1 && i + 1 >= ternary.size()) ternary.push_back(0);
         if (ternary[i] == 3) {
```
looks not tidy, but may suits compiler optimization

sfc9982 commented:
@BrandonPacewic It's done.

BrandonPacewic commented on src/balanced_ternary_converter.cpp:
```diff
@@ -28,25 +28,23 @@ std::vector<int> make_ternary(const std::vector<int>& values, int value) {
 void convert_to_balanced(std::vector<int>& ternary) {
-    for (int i = ternary.size() - 1; i >= 0; --i) {
+    for (int i = 0; i < ternary.size(); ++i) {
+        if (ternary[i] > 1 && i + 1 >= ternary.size()) ternary.push_back(0);
```
```suggestion
        if (ternary[i] > 1 && i + 1 >= ternary.size()) {
            ternary.push_back(0);
        }
```

I always prefer to have brackets, keeps things easy to read.

sfc9982 committed:
have brackets, keeps things easy to read.

Co-authored-by: Brandon P. <pacewicbrandon@gmail.com>

sfc9982 reviewed (commented):
I think it's okay to merge now.

BrandonPacewic reviewed (approved):
Looks good to me! Thanks for taking your time to correct this old project of mine 🎆

sfc9982 commented:
> Looks good to me! Thanks for taking your time to correct this old project of mine ![🎆](https://cdn.jsdelivr.net/gh/twitter/twemoji@14.0.2/assets/svg/1f386.svg)

You are welcome, glad to meet you!

sfc9982 commented:
> Ah I see my error now. Is there possibly a solution other than using `.insert` or `push_back`? As you said, particularly `.insert`, can be quite slow.

I spent some of my time on this.

First, try:

```cpp
#include <algorithm>
#include <chrono>
#include <cstring>
#include <iostream>
#include <vector>

using namespace std;

const int M = 500000;

void test_push_back() {
    vector<int> v;
    v.reserve(M);
    auto s = chrono::steady_clock::now();
    for (int i = 0; i < M; ++i)
        v.push_back(i);
    reverse(v.begin(), v.end());
    reverse(v.begin(), v.end());
    auto e = chrono::steady_clock::now();
    auto t = chrono::duration_cast<chrono::duration<double>>(e - s);
    cout << "push_back " << M << " integers cost " << t.count() << " seconds.\n";
}
void test_insert() {
    vector<int> v;
    v.reserve(M);
    auto s = chrono::steady_clock::now();
    for (int i = 0; i < M; ++i)
        v.insert(v.begin(), i);
    auto e = chrono::steady_clock::now();
    auto t = chrono::duration_cast<chrono::duration<double>>(e - s);
    cout << "insert front " << M << " integers cost " << t.count() << " seconds.\n";
}
void test_movebackward() {
    int *p = new int[M];
    auto s = chrono::steady_clock::now();
    for (int i = 0; i < M; ++i) {
        move_backward(p, p + i, p + i + 1);
        p[0] = i;
    }
    auto e = chrono::steady_clock::now();
    auto t = chrono::duration_cast<chrono::duration<double>>(e - s);
    cout << "insert front using move_backward " << M << " integers cost " << t.count() << " seconds.\n";
    delete[] p;
}
void test_byyourself() {
    int *p = new int[M];
    auto s = chrono::steady_clock::now();
    for (int i = 0; i < M; ++i) {
        int *first = p, *last = p + i, *result = p + i + 1;
        while (first != last)
            *--result = *--last;
        p[0] = i;
    }
    auto e = chrono::steady_clock::now();
    auto t = chrono::duration_cast<chrono::duration<double>>(e - s);
    cout << "insert front by yourself " << M << " integers cost " << t.count() << " seconds.\n";
    delete[] p;
}

int main() {
    test_push_back();
    test_insert();
    test_movebackward();
    test_byyourself();
    return 0;
}
```

which gives these on my computer with i7-8750H with no "-O" attached:

```
push_back 500000 integers cost 0.0227212 seconds.
insert front 500000 integers cost 15.5317 seconds.
insert front using move_backward 500000 integers cost 17.6967 seconds.
insert front by yourself 500000 integers cost 214.517 seconds.
```

From _godbolt_, I could see the two neighboring reverses are not optimized out. And it runs really fast. Through compiler optimization, it may be **O(1)**.

And the result shows use `.insert` is a really bad idea, and **insert without STL** is a worse idea.

`move_backward` will call some very good functions

In Windows, GCC compiles, you can see that it calls `msvcrt.memcpy`

In the assembly code, the compiler will be filled with NOP to complete the memory where ASM Command is located.

And they use [movups](https://www.felixcloutier.com/x86/movups) and [movntps](https://www.felixcloutier.com/x86/movntps)

It is an efficient transmission code that transmits 16 bytes at a time.

And [prefetchnta](https://www.felixcloutier.com/x86/prefetchh)

It's like put the data directly on the CPU face.

Add two optimizations as above and use four XMM registers to expand, then the operation of this memory is definitely fast.

BrandonPacewic commented:
It's quite interesting to observe the differences in time across different versions of the STL. With MSVC I am getting the following numbers with my Ryzen 5 4500U.

```
push_back 500000 integers cost 0.0463132 seconds.
insert front 500000 integers cost 13.2822 seconds.
insert front using move_backward 500000 integers cost 13.3125 seconds.
insert front by yourself 500000 integers cost 170.512 seconds.
```
